Add image column to resource pages

Ryex opened the pull request:
- Fixes exiting bug where simple selecting a resource would toggle it's enabled checkbox
- Load extra mod info from disk (image/icon/logo path, license, issueTracker) into `ModDetails`
- Add image column to resource pages for Mods, Resource Packs and Texture Packs
- Add rightclick menu to hide columns -> state saved to settings per page type
- Fix `PixmapCache` so that it can track cache misses from evictions (marked externally) and autogrow if they happen too fast.
- data providers can now set column resize modes

Fixes #1197
Fixes #356

Ryex committed:
feat(mod parsing): load extra mod details
- (image, license, issuetracker)

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(modpage): mod icon in description and column

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(texturepackPage): icon column

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(resourcePackPage): icon column

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: grow pixmapcache if it is evicting too fast

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

TayouVR commented on launcher/MTPixmapCache.h:
```diff
@@ -90,6 +94,43 @@ class PixmapCache final : public QObject {
     }
 
+    /** 
+     *  Mark that a cach miss occured because of a eviciton if too man of these occure to fast the cache size is increased 
```
```suggestion
     *  Mark that a cache miss occurred because of a eviction if too many of these occur too fast the cache size is increased 
```

TayouVR commented on launcher/minecraft/mod/Mod.cpp:
```diff
@@ -210,6 +215,47 @@ auto Mod::provider() const -> std::optional<QString>
+        qDebug() << "Mod" << name() << "Had it's icon evicted form the cache. reloading...";
+        PixmapCache::markCacheMissByEviciton();
+    }
+    // Imaged got evicted from the cache or an attmept to load it has not been made. load it and retry.
```
```suggestion
    // Image got evicted from the cache or an attempt to load it has not been made. load it and retry.
```

Ryex committed:
fix: fix qchar conversion and codeql

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
use 32x32 images for image column

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
change: don't toggle a resource's enabeling just by selecting it. only if they are on the checkbox.

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: display license and issue tracker

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
Added License and issuetracker url
![](https://cdn.discordapp.com/attachments/1031823065937629267/1104111113135734824/Screenshot_20230505_112311.png)
![](https://cdn.discordapp.com/attachments/1031823065937629267/1104148350682546326/Screenshot_20230505_135055.png)

Fabric of course uses sense-able square icons
![image](https://user-images.githubusercontent.com/508861/236577050-2352c4bc-6eb2-4956-8e57-c845029ae9e1.png)

Poopooracoocoo commented:
Are users able to right click on the column headers and show/hide columns? Likely out of scope though

embeddedt commented:
The row heights being uneven in the ATM8 screenshot bothers me from a design perspective.

Ryex commented:
> The row heights being uneven in the ATM8 screenshot bothers me from a design perspective.

sadly not something that could reasonably be addressed in this PR. it's a fundamental limitation of how the UI is set up for all the pages of that dialog.

Ryex committed:
feat: Column on left, hideable

- columns are hideable (saves to settings)
- image column moved to left
- datamodals can provide resize modes

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
UPDATE: 
- Added a rightclick hide menu for columns -> state saved to settings
- moved image column to the left
- resizemodes are settable by data providers
VIDEO: [PrismImageColumn.webm](https://github.com/PrismLauncher/PrismLauncher/assets/508861/9d55f99d-05ae-4bfa-8949-085082924b83)
![image](https://github.com/PrismLauncher/PrismLauncher/assets/508861/55809ec2-64f5-406f-8e51-6291e4bbeca4)

Ryex commented:
> Are users able to right click on the column headers and show/hide columns? Likely out of scope though

They can now!

flowln commented on launcher/minecraft/mod/ResourceFolderModel.cpp:
```diff
@@ -500,6 +505,67 @@ QVariant ResourceFolderModel::headerData(int section, Qt::Orientation orientatio
+        APPLICATION->settings()->getSetting(setting_name) : APPLICATION->settings()->registerSetting(setting_name);
+
+    auto hiddenColumns = QVariantUtils::toList<int>(setting->get());
+    auto index = hiddenColumns.indexOf(column);
```
I think a better saving strategy is to save the (not translated) column's header text, since saving the index will mean it will get invalid if we change ordering or add/remove columns in the future. If we use the text, then if we change a column's header text in the future, it will just stop being hidden, and I don't think we'll change one column's text to another's text, which would cause issues.

Ryex replied:
done

flowln commented on launcher/minecraft/mod/ResourceFolderModel.cpp:
```diff
@@ -500,6 +505,67 @@ QVariant ResourceFolderModel::headerData(int section, Qt::Orientation orientatio
+
+void ResourceFolderModel::saveHiddenColumn(int column, bool hidden)
+{
+    auto const setting_name = QString("UI/%1_Page/HiddenColumns").arg(id());
```
shouldn't this use the settings object of the instance associated with this model?

Ryex replied:
Should the setting really be instance specific? I suppose it could be but I was thinking it would be better as a application level setting.

Ryex committed:
Merge branch 'develop' into feature/images-for-resource-page

Ryex committed:
refactor: column names as class property, use string names in setting

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: use instance settings, make image column user resizeable, fix memory leak

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
- image col is now user re-sizeable (I feel this is the best way to handle the different width images)
- use instance settings so column hidden sittings are per instance
- fix memory leak (I'm just leaving address sanitizer on in my dev builds for now on)

TheKodeToad commented:
> > Are users able to right click on the column headers and show/hide columns? Likely out of scope though
> 
> They can now!

:sob: I was just trying to implement that and I completely forgot you did

DioEgizio commented:
![image](https://github.com/PrismLauncher/PrismLauncher/assets/83089242/aace70a2-853a-4447-aa40-95ea6db7957d)

the enable column is just too big. It doesn't seem too much of an issue until you notice what happens without icons

![image](https://github.com/PrismLauncher/PrismLauncher/assets/83089242/c937fc27-76d8-4d51-9302-370e440d2ea9)

Ryex committed:
fix: resize columns on hide ^& uniform heights

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
> ![image](https://user-images.githubusercontent.com/83089242/247206465-aace70a2-853a-4447-aa40-95ea6db7957d.png)
> 
> the enable column is just too big. It doesn't seem too much of an issue until you notice what happens without icons
> 
> ![image](https://user-images.githubusercontent.com/83089242/247207005-c937fc27-76d8-4d51-9302-370e440d2ea9.png)

fixed.

Trial97 commented:
This looks like it also fixes https://github.com/PrismLauncher/PrismLauncher/issues/356
